nvflare/edge/emulator/device_emulator.py

```diff
@@ -0,0 +1,111 @@
+from nvflare.edge.web.models.task_response import TaskResponse
+from nvflare.edge.web.models.user_info import UserInfo
+
+log = logging.getLogger(__name__)
+
+
+class DeviceEmulator:
+
+    def __init__(
+        self,
+        endpoint: str,
+        device_info: DeviceInfo,
+        user_info: UserInfo,
+        capabilities: Optional[dict],
+        processor: DeviceTaskProcessor,
+    ):
+        self.device_info = device_info
+        self.device_id = device_info.device_id
+        self.user_info = user_info
+        self.capabilities = capabilities
+        self.processor = processor
+        self.eta_api = EtaApi(endpoint, device_info, user_info)
+
+    def run(self):
+
+        try:
+
+            job = self.fetch_job()
+            self.processor.setup(job)
+            log.info(f"Received job: {job}")
+
+            while True:
+                task = self.fetch_task(job)
+                if not task:
+                    log.info(f"Job {job.job_Id} is done")
+                    break
+                log.info(f"Device:{self.device_id} Received task: {task}")
+
+                # Catch exception
+                result = self.processor.process_task(task)
+                log.info(f"Device:{self.device_id} Task processed. Result: {result}")
+                # Check result
+                result_response = self.eta_api.report_result(task, result)
+                log.info(f"Device:{self.device_id} Received result response: {result_response}")
+                task_done = task.get("task_done", False)
+                if task_done or result_response.status == "DONE":
+                    log.info(f"Job {job.job_id} {job.job_name} is done")
+                    break
+                elif result_response.status != "OK":
+                    log.error(f"Device:{self.device_id} Result report for task {task.task_name} is invalid")
+                    continue
+
+                log.info(f"Device:{self.device_id} Task {task.task_name} result reported successfully")
+
+            self.processor.shutdown()
+            log.info(f"Device:{self.device_id} Job {job.job_name} run ended")
+
+        except ApiError as error:
+            log.error(f"Status: {error.status}\nMessage: {str(error)}\nDetails: {error.details}")
+
+    def fetch_job(self) -> JobResponse:
+
+        while True:
+            job = self.eta_api.get_job(self.capabilities)
+            if job.status == "OK":
+                return job
+            elif job.status == "DONE":
+                job["job_done"] = True
+                return job
+            if job.status == "RETRY":
+                wait = job.retry_wait if job.retry_wait else 5
+                log.info(f"Device:{self.device_id} Retrying getting job in {wait} seconds")
+                time.sleep(wait)
+
+    def fetch_task(self, job: JobResponse) -> TaskResponse:
+
+        while True:
+            task = self.eta_api.get_task(job)
+            if task.status == "OK":
```

Comment:
Should define symbols for such constants (OK, DONE, RETRY, etc.) since they are used in multiple places.

Reply:
Added ApiStatus.